### PR TITLE
feat(metrics): Mock implementation for string indexer service

### PR DESCRIPTION
**Pull request by jjbayer**
Base class and mock implementation for the metrics string indexer service.

https://getsentry.atlassian.net/browse/INGEST-276

### CODE REVIEW

File: src/sentry/sentry_metrics/indexer/mock.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
+from typing import List, Optional, Tuple
+
+from sentry.models import Project
+
+from .base import StringIndexer, UseCase
+
+_STRINGS = {
+    "abnormal": 0,
```
Comment (untitaker) [inline src/sentry/sentry_metrics/indexer/mock.py:8]: do we want those numbers to be the respective FMV-hash already? I fear that if we ever switch to regular string indexer we will have to special-case those values forever

**Review by untitaker — APPROVED** [review]

File: src/sentry/sentry_metrics/indexer/base.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+class StringIndexer(Service):  # type: ignore
+    """
+    Provides integer IDs for metric names, tag keys and tag values
+    and the corresponding reverse lookup.
+    """
+
+    __all__ = ("record", "resolve", "reverse_resolve")
+
+    def record(self, project: Project, use_case: UseCase, string: str) -> int:
```
Comment (fpacifici) [inline src/sentry/sentry_metrics/indexer/base.py:22]: I wonder if this should be defined by project or by organization. 
Imagine a metric id. If we partition by project, the same metric (string) will have different numeric ids in different projects. If this is fine, then partitioning by project id is good. Can there be any case where this is not true ?

File: src/sentry/sentry_metrics/indexer/mock.py (hunk 1 above)
Comment (fpacifici) [reply src/sentry/sentry_metrics/indexer/mock.py:8]: We switched to a regular string indexer. We are not planning to use hashes anymore. The reason is that, in order to have a good confidence of no collisions, we had to partition the index in ways that made too many assumptions about how the product works that we cannot verify now.

File: src/sentry/sentry_metrics/indexer/base.py (hunk 2 above)
Comment (jan-auer) [reply src/sentry/sentry_metrics/indexer/base.py]: Agreed, we have this case in Release Health even. Updated.

**Review by dashed — APPROVED** [review]

**Pull request merged**